baseband/guppi/header.py

```diff
@@ -0,0 +1,375 @@
+        """
+        return (1. / self['TBIN']) * u.Hz
+
+    @sample_rate.setter
+    def sample_rate(self, sample_rate):
+        self['TBIN'] = 1. / abs(sample_rate.to_value(u.Hz))
+        bw = (sample_rate.to_value(u.MHz) * self['OBSNCHAN'] /
+              (1 if self.complex_data else 2))
+        self['OBSBW'] = (-1 if self.get('OBSBW', bw) < 0 else 1) * bw
+
+    @property
+    def sideband(self):
+        """True if upper sideband."""
+        return self['OBSBW'] > 0
+
+    @sideband.setter
+    def sideband(self, sideband):
+        self['OBSBW'] = (1 if sideband else -1) * abs(self['OBSBW'])
+
+    @property
+    def time_ordered(self):
+        return self['PKTFMT'] != 'SIMPLE'
+
+    @property
+    def samples_per_frame(self):
+        """Number of complete samples in the frame, including overlap."""
+        return self.payload_nbytes * 8 // self._bpcs
+
+    @samples_per_frame.setter
+    def samples_per_frame(self, samples_per_frame):
+        self.payload_nbytes = (samples_per_frame * self._bpcs + 7) // 8
+
+    @property
+    def overlap(self):
+        """Number of complete samples that overlap with the next frame."""
+        return self['OVERLAP']
+
+    @overlap.setter
+    def overlap(self, overlap):
+        self['OVERLAP'] = overlap.__index__()
+
+    @property
+    def offset(self):
+        """Offset from start of observation in units of time."""
+        # PKTIDX only counts valid packets, not overlap ones.
+        return self['STT_OFFS'] + ((self['PKTIDX'] * self['PKTSIZE'] * 8 //
+                                    self._bpcs) * self['TBIN'] * u.s)
+
+    @offset.setter
+    def offset(self, offset):
+        self['PKTIDX'] = int(round((offset.to_value(u.s) / self['TBIN'] /
+                                    self['PKTSIZE']) *
+                                   ((self._bpcs + 7) // 8)))
+
+    @property
+    def start_time(self):
+        """Start time of the observation."""
+        return (Time(self['STT_IMJD'], scale='utc', format='mjd') +
+                self['STT_SMJD'] * u.s)
+
+    @start_time.setter
+    def start_time(self, start_time):
+        start_time = Time(start_time, scale='utc', format='isot', precision=9)
+        self['STT_IMJD'] = int(start_time.mjd)
+        self['STT_SMJD'] = int(np.round(
+            (start_time - Time(self['STT_IMJD'], format='mjd',
+                               scale=start_time.scale)).sec))
+
+    @property
+    def time(self):
+        """Start time of the part of the observation covered by this header."""
+        return self.start_time + self.offset
+
+    @time.setter
+    def time(self, time):
+        """Set header time.
+
+        Note that this sets the start time of the header, assuming the offset
+        is already set correctly.
+
+        Parameters
+        ----------
+        time : `~astropy.time.Time`
+            Time for the first sample associated with this header.
+        """
+        self.start_time = time - self.offset
```

Comment:
Agreed, and changed.

Reply:
Actually, on second though, because this is identical to what DADA does, I think I'll keep it as is for now and we can raise an issue about it.

The trouble with setting `offset` rather than `start_time` is that for most other formats we only define a `time` when writing; this is also what our DADA examples do.  Since `offset` is by default zero, passing `time` to a DADA or GUPPI writer properly sets the start time, but setting `offset` in the setter for `time` means we'd need something like:

```
if 'STT_IMJD' not in self.keys():
    self.ooffset = 0
    self.start_time = time
else:
    self.offset = time - self.start_time
```

For DADA we'd replace `STT_IMJD` with `UTC_START`.

Reply:
OK, good to keep for now (and good to raise an issue so we can think about it)